### DISCUSSION TITLE
Adding zlib-ng as a third-party library

### DISCUSSION THREAD

**Discussion by kbieniex** [zlib-ng/zlib-ng#1367] [Q&A] [answered]
I'm trying to replace zlib with zlib-ng, as a third-party library to my project.
In my project's CMakeLists.txt:
`add_custom_target(zlib-ng)
set(ZLIB_COMPAT ON)
add_subdirectory("src/thirdparty/zlib-ng")`

But I'm getting an error during building: error: ‘Z_EXTERN’ does not name a type; did you mean ‘ZEXTERN’

Am I missing a step in CMake?

**Comment by mtl1979** [comment]
`add_subdirectory()` requires two parameters, so you are missing the second one.

Even though cmake documentation says the second parameter is optional, it only is optional when building in-tree, as when building out of source tree, the path in second parameter is used to find any generated files. The path in second parameter has to be subdirectory of current build directory.

**Comment by nmoinvaz** [comment]
I have an example on how to do this on the Wiki.

**Reply by kbieniex** [reply]
I followed the guide "Using zlib-ng with third-party projects in CMake" and my CMakeLists.tx looks like this:
`set(ZLIB_COMPAT ON)`
`add_subdirectory("src/thirdparty/zlib-ng" "src/thirdparty/zlib-ng" EXCLUDE_FROM_ALL)`

`SET(ZLIB_INCLUDE_DIR ${CMAKE_BINARY_DIR}/zlib-ng CACHE STRING "Path to zlib include directory")`
`SET(ZLIB_LIBRARY ZLIB:ZLIB CACHE STRING "Path to zlib directory")`
`include_directories(${ZLIB_INCLUDE_DIR})`

`add_library(ZLIB:ZLIB ALIAS zlib)`

I'm getting an error in one of my .cpp files that includes "zlib.h":
` error: ‘Bytef’ was not declared in this scope`

**Comment by mtl1979** [comment, marked as answer]
`Byfef` type was used in 16-bit processors, it has no use in 32-bit or 64-bit processors.

**Reply by kbieniex** [reply]
When I change it to Byte, same error. It seams that during build zconf is not seen by the compiler

**Reply by mtl1979** [reply]
That usually means that the second parameter to `add_subdirectory()` is not absolute path, it should start with contents of `CMAKE_CURRENT_BINARY_DIR`.

**Reply by kbieniex** [reply]
Changed it to "${CMAKE_BINARY_DIR}/src/thirdparty/zlib-ng", but still persists

**Reply by mtl1979** [reply]
Then you need to check that the include path in generated Makefiles match the location of `zconf.h`... As long as `ZLIB_INCLUDE_DIR` is set correctly and used, it should find the file. Looking at the earlier snippet, it is set incorrectly. It needs to include both the source and binary directories.

**Reply by kbieniex** [reply]
Solved it. `include_directories(${ZLIB_SOURCE_DIR})` also needs to be included, because I had a different subproject that had it's own zlib.h

**Comment by Shosh18** [comment]
I’m trying to do this on my iPad. I’m looking for the old a-library I can’t find it
